File: NTPClient.java
import java.io.IOException;
import java.net.*;
import java.nio.ByteBuffer;
import java.util.Date;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class NTPClient {

    private static final String LOG_FILE = "other_log.csv";
    private static final String[] NTD_IP = {"172.16.26.3", "172.16.26.4", "172.16.26.7", "172.16.26.9", "172.17.26.10", "172.16.26.11", "172.16.26.12", "172.16.26.13", "172.16.26.14", "172.16.26.15", "172.17.26.16", "172.17.26.17"};
    private static final int SYNC_TIME = 60; // in seconds

    private static Socket tcpClient;
    private static ExecutorService threadPool;
    private static long timestamp;
    private static int openNtdCount;
    private static int bias;


    public static void main(String[] args) {
        start();
    }

    private static void start() {
        String ntpServerName = "time.nplindia.org"; // Replace with the desired NTP server
        bias = 0; // Replace with the desired bias

        System.out.print("Synchronizing NTD:-\n");
        for (String host : NTD_IP) {
            System.out.println(host + "\n");
        }

        try {
            loop(ntpServerName);
        } catch (Exception e) {
            e.printStackTrace();
            System.out.println("Alert: Unable to get NTP Time...");
            try {
                Thread.sleep(60000); // Sleep for 1 minute
            } catch (InterruptedException ex) {
                ex.printStackTrace();
            }
            System.out.println("RESTARTING...");
            start();
        }
    }

    private static void loop(String ntpServerName) throws Exception {
        while (true) {
            try {
                openNtdCount = 0;
                threadPool = Executors.newFixedThreadPool(NTD_IP.length);
                syncNtd(ntpServerName);
                while (openNtdCount > 0) {
                    Thread.sleep(1000);
                }
                threadPool.shutdown();
            } finally {
                if (tcpClient != null && !tcpClient.isClosed()) {
                    tcpClient.close();
                }
            }
            Thread.sleep(SYNC_TIME * 1000); // Sleep for SYNC_TIME seconds
        }
    }

    private static void syncNtd(String server) throws Exception {
        System.out.println("Please wait while getting time from " + server + "\r\n");
        NTPStats response = request(server);
        timestamp = Math.round(response.getTxTimestamp() + bias);
        System.out.println("NPLI NTP TIME: " + new Date(timestamp * 1000).toString() + "\r\n");

        for (String host : NTD_IP) {
            openNtdCount++;
            threadPool.execute(() -> {
                try {
                    sendTime(host, timestamp, server);
                } catch (IOException e) {
                    e.printStackTrace();
                }
            });
        }
    }

    private static NTPStats request(String host) throws IOException {
        InetAddress address = InetAddress.getByName(host);
        DatagramSocket socket = new DatagramSocket();

        try {
            socket.setSoTimeout(5000); // Set socket timeout to 5 seconds

            NTPPacket queryPacket = new NTPPacket();
            queryPacket.setMode((byte) 3); // Mode 3 is client
            queryPacket.setVersion((byte) 3);


            DatagramPacket requestPacket = new DatagramPacket(queryPacket.toBytes(), queryPacket.toBytes().length, address, 123);
            socket.send(requestPacket);

            DatagramPacket responsePacket = new DatagramPacket(new byte[256], 256);
            socket.receive(responsePacket);

            NTPStats stats = new NTPStats();
            stats.fromBytes(responsePacket.getData());

            return stats;
        } finally {
            socket.close();
        }
    }

    private static void sendTime(String host, long timestamp, String server) throws IOException {
        String hostIp = host;
        int serverPort = 10000;
        tcpClient = new Socket(hostIp, serverPort);

        try {
            ByteBuffer buffer = ByteBuffer.allocate(28);
            buffer.put((byte) 0x55).put((byte) 0xAA).putInt(1).putInt(1).put((byte) 0xC1).putInt(15).putInt(15).put((byte) 0x10);
            buffer.putLong(timestamp * 1000); // Convert to milliseconds
            buffer.put((byte) 0).put((byte) 0).put((byte) 0x0D).put((byte) 0x0A);

            tcpClient.getOutputStream().write(buffer.array());

            byte[] received = new byte[1024];
            tcpClient.getInputStream().read(received);
            System.out.println(new Date(timestamp * 1000).toString() + ", " + host + ", Synchronized, " + bias + "\r\n");
        } catch (IOException e) {
            System.out.println(new Date(timestamp * 1000).toString() + ", " + host + ", Not Connected, " + bias + "\r\n");
        } finally {
            openNtdCount--;
            tcpClient.close();
        }
    }
}

class NTPStats extends NTPPacket {
    private long destTimestamp;

    // Constructors, getters, and setters omitted for brevity


    public void setDestTimestamp(long destTimestamp) {
        this.destTimestamp = destTimestamp;
    }
}

File: ntp_client.py
# ntp_client.py

import socket
import time
import datetime
from ntp_packet import NTPPacket
from ntp_stats import NTPStats, ntp_to_system_time, NTPException
class NTPClient:
    """NTP client session."""

    def __init__(self):
        """Constructor."""
        pass

    def request(self, host, version=2, port='ntp', timeout=5):
        """Query a NTP server.

        Parameters:
        host    -- server name/address
        version -- NTP version to use
        port    -- server port
        timeout -- timeout on socket operations

        Returns:
        NTPStats object
        """
        # lookup server address
        addrinfo = socket.getaddrinfo(host, port)[0]
        family, sockaddr = addrinfo[0], addrinfo[4]

        # create the socket
        s = socket.socket(family, socket.SOCK_DGRAM)

        try:
            s.settimeout(timeout)

            # create the request packet - mode 3 is client
            query_packet = NTPPacket(mode=3, version=version,
                                tx_timestamp=system_to_ntp_time(time.time()))

            # send the request
            s.sendto(query_packet.to_data(), sockaddr)

            # wait for the response - check the source address
            src_addr = None,
            while src_addr[0] != sockaddr[0]:
                response_packet, src_addr = s.recvfrom(256)

            # build the destination timestamp
            dest_timestamp = system_to_ntp_time(time.time())
        except socket.timeout:
            raise NTPException("No response received from %s." % host)
        finally:
            s.close()

        # construct corresponding statistics
        stats = NTPStats()
        stats.from_data(response_packet)
        stats.dest_timestamp = dest_timestamp

        return stats
    
def system_to_ntp_time(timestamp):
    """Convert a system time to a NTP time.

    Parameters:
    timestamp -- timestamp in system time

    Returns:
    corresponding NTP time
    """
    return timestamp + NTP.NTP_DELTA


